File: cytof/play/vb/wikipedia_example/vb.R
vb <- function(init, update, compare, eps=1E-6) {
  param <- init

  its <- 0
  converged <- FALSE

  while (!converged) {
    its <- its + 1
    updt <- update(param)
    converged <- all(compare(updt, param) < eps)
    param <- updt
  }

  list(param=param, its=its)
}

File: cytof/play/vb/wikipedia_example/vb.scala
def vb[S](init:S, update:S=>S, compare:(S,S)=>Double, eps:Double=1E-6) = {

  def findOpt(curr:S, its:Int): (S,Int) = {
    val next = update(curr)
    if (math.abs(compare(next, curr)) < eps) 
      (next, its)
    else
      findOpt(next, its + 1)
  }

  findOpt(init, 0)
}
